File: 프로그래머스/3/43162. 네트워크/네트워크.cpp
#include <string>
#include <vector>
#include <iostream>
#include <deque>

using namespace std;
vector<bool> visited;

void bfs(vector<vector<int>> computers, int start){
    deque<int> dq;
    dq.push_back(start);
    
    while (!dq.empty()){
        int curr = dq.front();
        dq.pop_front();
        vector<int> connected = computers[curr];
        for (int c = 0; c < connected.size(); c++){
            if (connected[c] == 1 && !visited[c]){
                dq.push_back(c);
                visited[c] = true;
            }
        }
        
    }
}


int solution(int n, vector<vector<int>> computers) {
    int answer = 0;
    for (int i = 0; i < n; i++){
        visited.push_back(false);
    }
    
    for (int c = 0; c < n; c++){
        if (!visited[c]){
            visited[c] = true;
            bfs(computers, c);
            answer ++;
        }
    }
    
    return answer;
}

File: 프로그래머스/3/43162. 네트워크/네트워크.java
import java.util.*;


class Solution {
    public int solution(int n, int[][] computers) {
        int answer = 0;
        int[] visited = new int[n];
        
        for (int i = 0; i < n; i++){
            // System.out.println("start: " + i);
            Queue<Integer> dq = new LinkedList<>();
            
            if (visited[i] == 0){
                dq.offer(i);
                visited[i] = 1;
            
                while (!dq.isEmpty()){
                    int curr = dq.poll();
                    // System.out.println("curr: " + curr);

                    for (int j = 0; j < n; j++){
                        if(visited[j] == 0 && computers[curr][j] == 1){
                            dq.offer(j);
                            visited[j] = 1;
                        }
                    }
                }
                answer ++;
            }
        }
        
        return answer;
    }
}